--- a/arch/mips/bcm63xx/cpu.c
+++ b/arch/mips/bcm63xx/cpu.c
@@ -317,7 +317,7 @@ void __init bcm63xx_cpu_init(void)
 			bcm63xx_irqs = bcm6358_irqs;
 		} else {
 			/* all newer chips have the same chip id location */
-			u16 chip_id = bcm_readw(BCM_6368_PERF_BASE);
+			u16 chip_id = bcm_readw(BCM_6368_PERF_BASE) & ~1;
 
 			switch (chip_id) {
 			case BCM6328_CPU_ID:
@@ -356,6 +356,10 @@ void __init bcm63xx_cpu_init(void)
 	bcm63xx_cpu_id = (tmp & REV_CHIPID_MASK) >> REV_CHIPID_SHIFT;
 	bcm63xx_cpu_rev = (tmp & REV_REVID_MASK) >> REV_REVID_SHIFT;
 
+	/* Properly detect certain CPUs */
+	if (bcm63xx_cpu_id != BCM6345_CPU_ID)
+		bcm63xx_cpu_id &= ~1;
+
 	if (bcm63xx_cpu_id != expected_cpu_id)
 		panic("bcm63xx CPU id mismatch");
 
